File: hard/2790-maximum-number-of-groups-with-increasing-length/solution.py
from typing import List


class Solution:
    def maxIncreasingGroups(self, usageLimits: List[int]) -> int:
        """
        2790. Maximum Number of Groups With Increasing Length
        Time: O(n log n)
        Space: O(1)
        """
        usageLimits.sort()

        running = 0
        groups = 0

        for limit in usageLimits:
            running += limit
            if running >= groups + 1:
                running -= groups + 1
                groups += 1

        return groups

File: hard/2790-maximum-number-of-groups-with-increasing-length/solution.java
import java.util.Arrays;

/**
 * 2790. Maximum Number of Groups With Increasing Length
 * Time: O(n log n)
 * Space: O(1)
 */
class Solution {
    public int maxIncreasingGroups(int[] usageLimits) {
        Arrays.sort(usageLimits);

        long running = 0;
        long groups = 0;

        for (int limit : usageLimits) {
            running += limit;
            if (running >= groups + 1) {
                running -= groups + 1;
                groups += 1;
            }
        }

        return (int) groups;
    }
}
